src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,375 @@
+assertEqual msg a b = if a == b
+  then return ()
+  else error $ msg ++ ": " ++ show a ++ " is not equal to " ++ show b
+
+_insertionTest  :: IO (Either Error ())
+_insertionTest = withTempSQLiteConnection $ \case
+  Left l -> return (Left $ fst l)
+  Right c -> do
+    createBlockHistoryTable c
+    nonVersionInsert c (BlockHistory (BlockHeight 0) nullBlockHash)
+    res1 <- qry_ c "SELECT * FROM BlockHistory;" [RInt, RBlob]
+    assertEqual "Read the BlockHistoryTable Table" res1 ans1
+    createVersionHistoryTable c
+    nonVersionInsert c (VersionHistory (BlockVersion (BlockHeight 0) (ReorgVersion 0)))
+    res2 <- qry_ c "SELECT * FROM VersionHistory;" [RInt, RInt]
+    assertEqual "Read the VersionHistoryTable Table" res2 ans2
+    createVersionedTablesTable c
+    nonVersionInsert c
+        (VersionedTables (TableName "user1") (BlockVersion (BlockHeight 0) (ReorgVersion 0)))
+    res3 <- qry_ c "SELECT * FROM VersionedTables;" [RText, RInt, RInt]
+    assertEqual "Read the VersionedTables Table" res3 ans3
+    return $ Right ()
+  where
+    ans1 =
+      [ [ SInt 0
+        , SBlob
+            "\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL\NUL"
+        ]
+      ]
+    ans2 = [[SInt 0,SInt 0]]
+    ans3 = [[SText "user1",SInt 0,SInt 0]]
+
+
+createBlockHistoryTable :: Database -> IO ()
+createBlockHistoryTable c =
+  exec_
+    c
+    "CREATE TABLE BlockHistory(blockheight UNSIGNED BIGINT PRIMARY KEY,hash BLOB);"
+
+createVersionHistoryTable  :: Database -> IO ()
+createVersionHistoryTable c =
+  exec_
+    c
+    "CREATE TABLE VersionHistory (version UNSIGNED BIGINT PRIMARY KEY,blockheight UNSIGNED BIGINT);"
+
+createVersionedTablesTable  :: Database -> IO ()
+createVersionedTablesTable c =
+  exec_
+    c
+    "CREATE TABLE VersionedTables (tableid TEXT PRIMARY KEY\
+  \ , blockheight UNSIGNED BIGINT\
+  \ , version UNSIGNED BIGINT);"
+
+{-
+
+for ints, investigate which int?
+
+  history of BH pairs  -- BlockHistory
+
+  schema BlockHistory
+
+  create table BlockHistory
+    (block  INT PRIMARY KEY
+    ,**header    varchar<256>) ** investigate best type for hashes
+    (BLOB seems to be the answer.)
+
+
+  history of BV pairs -- VersionHistory
+
+  schema VersionHistory
+
+  create table VersionHistory
+    (version  INT PRIMARY KEY
+    ,block    INT)
+
+  table management table -- VersionedTables
+
+  schema for VersionedTables
+
+  create table VersionedTables
+        (tableid   varchar<256> PRIMARY KEY
+        , block    INT
+        ,version   INT
+        , touchedBlock INT ** maybe
+        , touchedVersion INT ** maybe
+        )
+
+  tableid (primary key)
+
+  touched table
+
+
+
+-}
+
+detectVersionChange ::
+     Database
+  -> BlockHeight
+  -> BlockHash
+  -> IO (Either String OnVersionChange)
+detectVersionChange c bRestore hsh  = do
+  let sqlBtoBH [[SInt bh]] = BlockHeight (fromIntegral bh)
+      sqlBtoBH _ = error "sqlBtoBH: wat"
```

Comment:
This is a bad style, you should never `fmap` a conversion function that can fail. See below for more idiomatic code.

Reply:
Understood. Maybe I could use `expectSing` like in the suggestion below?

Reply:
My hunch is that it might be better to use throwDbError (from Pact.Types.Runtime) instead of die from System.Exit?